### PR TITLE
Optimize SM2 on aarch64

### PR DESCRIPTION
**Pull request by xu-yi-zhou**
This patch optimizes SM2 for ARM processor using A64 instruction and precomputation table, which can speed up SM2 sign about 10 times and SM2 verify about 3 times. A new configure option `no-sm2-precomp` has been added to disable the precomputed table for point multiplicatin of the base point.

Perf data on Kunpeng-920 2.6GHz hardware looks like this:

|        | sign    | verify  | sign/s  | verify/s |
|--------|---------|---------|---------|----------|
| Before | 0.0004s | 0.0004s | 2438.2  | 2460.9   |
| After  | 0.0000s | 0.0001s | 27137.4 | 7434.0   |
| After (no-sm2-precomp)  | 0.0001s   | 0.0002s   | 7745.5   | 4411.2|

Signed-off-by: Xu Yizhou [xuyizhou1@huawei.com](mailto:xuyizhou1@huawei.com)

##### Checklist

- [ ] documentation is added or updated

### CODE REVIEW

**Comment by zzl360** [comment]
any benchmark data?

File: crypto/ec/build.info [hunk 1]
```diff
@@ -104,3 +104,6 @@ GENERATE[ecp_nistp521-ppc64.s]=asm/ecp_nistp521-ppc64.pl
 
 GENERATE[x25519-x86_64.s]=asm/x25519-x86_64.pl
 GENERATE[x25519-ppc64.s]=asm/x25519-ppc64.pl
+
+GENERATE[ecp_sm2p256-armv8.S]=asm/ecp_sm2p256-armv8.pl
+INCLUDE[ecp_sm2p256-armv8.o]=..
```
Comment (paulidale) [inline crypto/ec/build.info:109]: Newline at end of file please.

File: crypto/ec/ecp_sm2p256.c [hunk 2]
```diff
@@ -0,0 +1,816 @@
...
+/* Modular sub: r = a-b mod n, where n = ord(p) */
+void ecp_sm2p256_sub_mod_ord(BN_ULONG *r, const BN_ULONG *a, const BN_ULONG *b);
+/* Modular mul by 3: out = 3*a mod p */
+void ecp_sm2p256_mul_by_3(BN_ULONG *r, const BN_ULONG *a);
+/* Modular mul: r = a*b mod p */
+void ecp_sm2p256_mul(BN_ULONG *r, const BN_ULONG *a, const BN_ULONG *b);
+/* Modular sqr: r = a^2 mod p */
+void ecp_sm2p256_sqr(BN_ULONG *r, const BN_ULONG *a);
+
+static ossl_inline BN_ULONG is_zero(BN_ULONG a)
```
Comment (paulidale) [inline crypto/ec/ecp_sm2p256.c:108]: These would be better added to internal constant time header I think.

File: crypto/ec/ecp_sm2p256.c [hunk 3]
```diff
@@ -0,0 +1,816 @@
...
+#define ecp_sm2p256_bignum_field_elem(out, in) bn_copy_words(out, in, P256_LIMBS)
+
+/* Binary algorithm for inversion in Fp */
+#define BN_MOD_INV(out, in, fadd, fsub, mod, mod_div_2) \
+{                                                       \
+        BN_ULONG u[4] ALIGN32;                          \
+        BN_ULONG v[4] ALIGN32;                          \
+        BN_ULONG x1[4] ALIGN32 = {1, 0, 0, 0};          \
+        BN_ULONG x2[4] ALIGN32 = {0};                   \
+        if (is_zeros(in))                               \
```
Comment (paulidale) [inline crypto/ec/ecp_sm2p256.c:171]: Blank line above.

**Review by paulidale — COMMENTED** [review]
Some nits.

That's quite a performance boost and one large file.

File: crypto/ec/ecp_sm2p256.c (hunk 2 above)
Comment (xu-yi-zhou) [reply crypto/ec/ecp_sm2p256.c]: I replaced `is_zero()` with `constant_time_is_zero_64()` which is defined in constant_time.h, is it OK?

**Comment by xu-yi-zhou** [comment]
Hi @InfoHunter, could you please review this patch?

**Comment by xu-yi-zhou** [comment]
ping for review

File: crypto/ec/ecp_sm2p256.c [hunk 4]
```diff
@@ -0,0 +1,810 @@
...
+                fsub(x1, x1, x2);                       \
+            } else {                                    \
+                fsub(v, v, u);                          \
+                fsub(x2, x2, x1);                       \
+            }                                           \
+        }                                               \
+        if (is_one(u))                                  \
+            memcpy(out, x1, 32);                        \
+        else                                            \
+            memcpy(out, x2, 32);                        \
```
Comment (t8m) [inline crypto/ec/ecp_sm2p256.c:199]: The indentation is wrong here - too much

**Review by t8m — CHANGES_REQUESTED** [review]
I wonder if this optimization should be excluded if OPENSSL_SMALL_FOOTPRINT is defined. Given the large table size.

File: crypto/ec/ecp_sm2p256.c (hunk 4 above)
Comment (xu-yi-zhou) [reply crypto/ec/ecp_sm2p256.c:199]: Sorry, could you please give me an example? Why the indentation is wrong here

Comment (t8m) [reply crypto/ec/ecp_sm2p256.c:199]: it is too far to the right. The base indent should be just 4 spaces.

Comment (xu-yi-zhou) [reply crypto/ec/ecp_sm2p256.c:199]: OK, I will fix it.

**Comment by paulidale** [comment]
> I wonder if this optimization should be excluded if OPENSSL_SMALL_FOOTPRINT is defined. Given the large table size.

It would be a bonus but aarch64 isn't normally associated with _small_ devices.

File: crypto/ec/asm/ecp_sm2p256-armv8.pl [hunk 5]
```diff
@@ -0,0 +1,843 @@
...
+       .globl  ecp_sm2p256_mul
+       .type   ecp_sm2p256_mul, %function
+       .align 6
+
+
+ecp_sm2p256_mul:
+       # Store scalar registers
+       stp     x29, x30, [sp, #-80]!
+    add     x29, sp, #0
+    stp     x16, x17, [sp, #16]
```
Comment (docularxu) [inline crypto/ec/asm/ecp_sm2p256-armv8.pl:525]: Indentation: better to align Line 525 and Line 524.
https://www.openssl.org/policies/technical/coding-style.html

File: crypto/ec/asm/ecp_sm2p256-armv8.pl [hunk 6]
```diff
@@ -0,0 +1,843 @@
...
+       adc      x15,xzr,xzr
+       mov             x11,x3
+       mov             x12,x4
+       mov             x13,x5
+       mov             x14,x6
+       /* Sub polynomial */
+       adr             x2,$mod
+       ldp             x7,x8,[x2]
+       ldp             x9,x10,[x2,#0x10]
+       subs    x11,x11,x7
```
Comment (docularxu) [inline crypto/ec/asm/ecp_sm2p256-armv8.pl:42]: would you please to keep these instruction operands aligned?
L39 ~ L41, etc.

**Comment by tom-cosgrove-arm** [comment]
(As I understand it, @docularxu is still reviewing, and there should be an update to the PR from the OP)

**Comment by tom-cosgrove-arm** [comment]
(removed the `stalled` label as the PR has been updated)

**Comment by docularxu** [comment]
I'm from Linaro and specializing in Arm platform optimization. The latest version from @xu-yi-zhou effectively addresses the concerns I had previously raised. Also, I verified this version on Apple silicon M2, and the results align with the performance improvements mentioned earlier:
> which can speed up SM2 sign about 10 times and SM2 verify about 3 times.

With these, I would recommend merging this patch into the master branch.

File: crypto/ec/build.info [hunk 7]
```diff
@@ -29,8 +29,9 @@ IF[{- !$disabled{asm} -}]
 
   $ECASM_armv4=ecp_nistz256.c ecp_nistz256-armv4.S
   $ECDEF_armv4=ECP_NISTZ256_ASM
-  $ECASM_aarch64=ecp_nistz256.c ecp_nistz256-armv8.S
-  $ECDEF_aarch64=ECP_NISTZ256_ASM
+  $ECASM_aarch64=ecp_nistz256.c ecp_nistz256-armv8.S ecp_sm2p256.c \
+                 ecp_sm2p256_table.c ecp_sm2p256-armv8.S
+  $ECDEF_aarch64=ECP_NISTZ256_ASM ECP_SM2P256_ASM
```
Comment (t8m) [inline crypto/ec/build.info:34]: This needs to be conditional on sm2 being enabled:

```suggestion
  $ECASM_aarch64=ecp_nistz256.c ecp_nistz256-armv8.S
  $ECDEF_aarch64=ECP_NISTZ256_ASM
  IF[{- !$disabled{'sm2'} -}]
    $ECASM_aarch64=$ECASM_aarch64ecp ecp_sm2p256.c \
                   ecp_sm2p256_table.c ecp_sm2p256-armv8.S
    $ECDEF_aarch64=$ECDEF_aarch64 ECP_SM2P256_ASM
  ENDIF
```

File: crypto/ec/build.info [hunk 8]
```diff
@@ -125,3 +126,6 @@ IF[{- !$disabled{'ecx'} -}]
 GENERATE[x25519-x86_64.s]=asm/x25519-x86_64.pl
 GENERATE[x25519-ppc64.s]=asm/x25519-ppc64.pl
 ENDIF
+
+GENERATE[ecp_sm2p256-armv8.S]=asm/ecp_sm2p256-armv8.pl
+INCLUDE[ecp_sm2p256-armv8.o]=..
```
Comment (t8m) [inline crypto/ec/build.info:131]: Similarly to above, make it conditional on sm2 being enabled.

**Review by t8m — CHANGES_REQUESTED** [review]

**Comment by xu-yi-zhou** [comment]
Hi all, 

I'm thinking about whether to implement constant-time point multiplication and modular inversion to avoid side channel attack.

I'm trying the following to fix side channel,
* implement constant-time `ecp_sm2p256_point_double`, `ecp_sm2p256_point_add_affine` and `ecp_sm2p256_point_add`
* ignore the check for `index` in  `ecp_sm2p256_point_G_mul_by_scalar` and `ecp_sm2p256_point_P_mul_by_scalar`

After these changes, the performance of `sign` is reduced by 2% (27137.4->26641.9) and the one of `verify` is reduced by 6% (7434.0->6973.9).

If the side channel protection is necessary, I still need to implement a `gather` method for precomputed table like `ecp_nistz256_gather_w7`, `ecp_nistz256_gather_w5` to avoid memory access patterns, a constant time `BN_MOD_INV` for modular inversion. I'm not sure how much performance will decrease.

Welcome to discsuss.

File: crypto/ec/build.info (hunk 7 above)
Comment (xu-yi-zhou) [reply crypto/ec/build.info]: fixed on https://github.com/openssl/openssl/pull/20754/commits/106d6f4b6a61d0e517eb1c6053e8222665baeb52

File: crypto/ec/build.info (hunk 8 above)
Comment (xu-yi-zhou) [reply crypto/ec/build.info:136]: fixed on https://github.com/openssl/openssl/commit/106d6f4b6a61d0e517eb1c6053e8222665baeb52

**Comment by InfoHunter** [comment]
> Hi all,
> 
> I'm thinking about whether to implement constant-time point multiplication and modular inversion to avoid side channel attack.
> 
> I'm trying the following to fix side channel,
> 
> * implement constant-time `ecp_sm2p256_point_double`, `ecp_sm2p256_point_add_affine` and `ecp_sm2p256_point_add`
> * ignore the check for `index` in  `ecp_sm2p256_point_G_mul_by_scalar` and `ecp_sm2p256_point_P_mul_by_scalar`
> 
> After these changes, the performance of `sign` is reduced by 2% (27137.4->26641.9) and the one of `verify` is reduced by 6% (7434.0->6973.9).
> 
> If the side channel protection is necessary, I still need to implement a `gather` method for precomputed table like `ecp_nistz256_gather_w7`, `ecp_nistz256_gather_w5` to avoid memory access patterns, a constant time `BN_MOD_INV` for modular inversion. I'm not sure how much performance will decrease.
> 
> Welcome to discsuss.

Yes, I consider anti-side-channel should be a 'must' stuff prior to performance.

**Comment by xu-yi-zhou** [comment]
Thank you all for your reply!

Currently, I have fixed all the review comments and look forward to the approvals.

File: crypto/ec/build.info [hunk 9]
```diff
@@ -125,3 +130,8 @@ IF[{- !$disabled{'ecx'} -}]
 GENERATE[x25519-x86_64.s]=asm/x25519-x86_64.pl
 GENERATE[x25519-ppc64.s]=asm/x25519-ppc64.pl
 ENDIF
+
+IF[{- !$disabled{'sm2'} -}]
+GENERATE[ecp_sm2p256-armv8.S]=asm/ecp_sm2p256-armv8.pl
```
Comment (paulidale) [inline crypto/ec/build.info:135]: We'd usually indent these two spaces.

File: crypto/ec/ecp_sm2p256.c [hunk 10]
```diff
@@ -0,0 +1,791 @@
...
+
+#include <string.h>
+#include <openssl/err.h>
+#include "crypto/bn.h"
+#include "ec_local.h"
+#include "internal/constant_time.h"
+
+#if defined(__GNUC__)
+#define ALIGN32 __attribute((aligned(32)))
+#define ALIGN64 __attribute((aligned(64)))
```
Comment (paulidale) [inline crypto/ec/ecp_sm2p256.c:25]: Add a space between the `#` and the `define` here and below.
Our coding style indents nested preprocessor directives.

File: crypto/ec/ecp_sm2p256.c [hunk 11]
```diff
@@ -0,0 +1,791 @@
...
+void ecp_sm2p256_sqr(BN_ULONG *r, const BN_ULONG *a);
+
+static ossl_inline BN_ULONG is_zeros(const BN_ULONG *a)
+{
+    BN_ULONG res;
+
+    res = a[0] ^ 0;
+    res |= a[1] ^ 0;
+    res |= a[2] ^ 0;
+    res |= a[3] ^ 0;
```
Comment (paulidale) [inline crypto/ec/ecp_sm2p256.c:106]: Why include the `^ 0` which seems like it does nothing?

File: crypto/ec/ecp_sm2p256.c [hunk 12]
```diff
@@ -0,0 +1,791 @@
...
+    ecp_sm2p256_mul(tmp0, tmp0, Q->X);
+    ecp_sm2p256_mul(tmp1, tmp1, Q->Y);
+    ecp_sm2p256_sub(tmp0, tmp0, P->X);
+    ecp_sm2p256_sub(tmp1, tmp1, P->Y);
+
+    /* zero-check tmp0, tmp1 */
+    if (is_zeros(tmp0)) {
+        if (is_zeros(tmp1)) {
+            P256_POINT K;
+            for (i = 0; i < P256_LIMBS; ++i) {
```
Comment (paulidale) [inline crypto/ec/ecp_sm2p256.c:265]: Blank line between declaration and code.

File: crypto/ec/ecp_sm2p256.c [hunk 13]
```diff
@@ -0,0 +1,791 @@
...
+
+            if ((tmp = BN_CTX_get(ctx)) == NULL)
+                goto err;
+            if (!BN_nnmod(tmp, scalar[i], group->order, ctx)) {
+                ECerr(ERR_LIB_EC, ERR_R_BN_LIB);
+                goto err;
+            }
+            scalars[i] = tmp;
+        } else
+            scalars[i] = scalar[i];
```
Comment (paulidale) [inline crypto/ec/ecp_sm2p256.c:547]: { } around the else case.

When either half of the if has {}, both should.

**Review by paulidale — CHANGES_REQUESTED** [review]
Just coding style nits.

File: crypto/ec/ecp_sm2p256.c [hunk 14]
```diff
@@ -0,0 +1,790 @@
...
+
+#include <string.h>
+#include <openssl/err.h>
+#include "crypto/bn.h"
+#include "ec_local.h"
+#include "internal/constant_time.h"
+
+#if defined(__GNUC__)
+# define ALIGN32 __attribute((aligned(32)))
+# define ALIGN64 __attribute((aligned(64)))
```
Comment (docularxu) [inline crypto/ec/ecp_sm2p256.c:25]: I noticed ALIGN32 are declared in two .c files. one is here. the other is in [L41 of crypto/ec/ecp_nistz256.c](https://github.com/openssl/openssl/blob/c612289b77c37f7295d5af0d0e6b6c04e6ba727c/crypto/ec/ecp_nistz256.c#L41)

Considering this as a common feature, should we put these macros into a common file? such as ec_local.h in crypto/ec/?

**Comment by docularxu** [comment]
The feedbacks/comments so far have been fixed by yizhou in the latest patchset.
I verified again on the following setup. This patchset effectively enhances the performance as for aarch64 platforms, as claimed by the committer. I saw Sign/s 10 times gain, Verify/s 2.x times gain.
 - OS : Ubuntu 22.04 (via Parallel VM)
 - CPU: aarch64 Apple M2 Pro
 - C Compiler:     Version 11.3.0 (at /usr/bin/gcc)
 - C++ Compiler:   Version 11.3.0 (at /usr/bin/g++)

With that, I would give my approval on this patch. Hope it helps maintainers to merge the PR.

File: crypto/ec/ecp_sm2p256.c (hunk 14 above)
Comment (paulidale) [reply crypto/ec/ecp_sm2p256.c:25]: They would be better in a header file.  `include/internal/e_os.h` seems like a reasonable place.

Comment (tom-cosgrove-arm) [reply crypto/ec/ecp_sm2p256.c:25]: Should this change be made as part of this pull request, or can it be done in another, later, one?

Comment (t8m) [reply crypto/ec/ecp_sm2p256.c:25]: I think doing it later would be OK, if someone really does it :grin:

Comment (tom-cosgrove-arm) [reply crypto/ec/ecp_sm2p256.c:25]: I _can_ do that!

So can we get approvals on this PR and get it merged?

**Comment by t8m** [comment]
I am still worried about the size of the precomputed table - its size is 512kBytes, that's definitely not something to ignore in regards to the size of the rest of libcrypto. My feeling is we should at least make it possible to disable adding this if desired - i.e. either respect OPENSSL_SMALL_FOOTPRINT or even add a completely new define to disable it.

**Comment by t8m** [comment]
> The size of the precomputed table in file `ecp_nistz256_table.c` is 148kBytes, and it is included in libcrypto by default and no define to disable it (correct me if I'm wrong). I was wondering why this precomputed table is acceptable, is it because the size or something else?

Well.. nobody asked for the define when that one was included. It is also significantly smaller than the SM2 precomputed table. However it would be of course nice to have that one disableable as well.

**Comment by t8m** [comment]
**OTC question: Is the 512 kB read-only data table acceptable as is or do we require a possibility to disable it?**

**Comment by paulidale** [comment]
I think that `OPENSSL_SMALL_FOOTPRINT` should apply here.

That doesn't preclude having another mechanism as well, `OPENSSL_SMALL_FOOTPRINT` impacts a number of other algorithms and areas (reducing space and performance).  I expect a lot of people don't use SM2 and wouldn't want to pay the `OPENSSL_SMALL_FOOTPRINT` costs.

**Comment by t8m** [comment]
**OTC: We need compile-time config option for this that is on by default. The increased size of libcrypto also needs to be documented in CHANGES.md.**

**Comment by t8m** [comment]
@xu-yi-zhou Could you please add a new compile-time config option for this as OTC requested? It would not be a define but a full option like `sm2` itself is. I.e., `sm2-precomp` or something like that.

**Comment by xu-yi-zhou** [comment]
> We need compile-time config option for this that is on by default. The increased size of libcrypto also needs to be documented in CHANGES.md.

 OK, I will do that as soon as possible.

**Comment by xu-yi-zhou** [comment]
Hi all, I'm trying to add an option `enable-sm2-precomp` to enable the precomputed table. Do I need to add this option in `.github/workflows/*.yml`?

**Comment by t8m** [comment]
Yes, please. Add `no-sm2-precomp` to run-checker-daily.yml. Although it won't test much as this is arm specific and run-checker-daily runs on x86_64, it will at least test the disable-ability of the option.

**Comment by tom-cosgrove-arm** [comment]
> OTC: We need compile-time config option for this that is on by default. The increased size of libcrypto also needs to be documented in CHANGES.md.

My understanding of this request is that the large table would be included in default AArch64 builds, but can be disabled for smaller builds by configuring with `no-sm2-precomp`.

This seems to only use the large table if `enable-sm2-precomp` is given - @t8m is that what you wanted?

**Comment by xu-yi-zhou** [comment]
@tom-cosgrove-arm I think you are right, and I also forgot to add the  increased size of libcrypto in CHANGES.md, I will fix these.

File: CHANGES.md [hunk 15]
```diff
@@ -25,6 +25,13 @@ OpenSSL 3.2
 
 ### Changes between 3.1 and 3.2 [xx XXX xxxx]
 
+ * Added optimization for SM2 algorithm on aarch64. It uses a huge precompted
+   table for point multiplicatin of the base point, which increases the size of
```
Comment (t8m) [inline CHANGES.md:29]: Typo: multiplication

File: INSTALL.md [hunk 16]
```diff
@@ -914,6 +914,10 @@ Do not create shared libraries, only static ones.
 
 See [Notes on shared libraries](#notes-on-shared-libraries) below.
 
+### no-sm2-precomp
+
+Don't support for the SM2 precomputed table on aarch64.
```
Comment (t8m) [inline INSTALL.md:919]: I'd suggest: `Disable using the SM2 precomputed table on aarch64 to make the library smaller.`

File: crypto/ec/build.info [hunk 17]
```diff
@@ -31,6 +31,13 @@ IF[{- !$disabled{asm} -}]
   $ECDEF_armv4=ECP_NISTZ256_ASM
   $ECASM_aarch64=ecp_nistz256.c ecp_nistz256-armv8.S
   $ECDEF_aarch64=ECP_NISTZ256_ASM
+  IF[{- !$disabled{'sm2'} -}]
+    $ECASM_aarch64=$ECASM_aarch64 ecp_sm2p256.c ecp_sm2p256-armv8.S
```
Comment (t8m) [inline crypto/ec/build.info:35]: Should ecp_sm2p256.c and  ecp_sm2p256-armv8.S be also conditional on `!$disabled{'sm2-precomp'}`? I assume these are unusable without the table.

I.e., there should be only a single `IF [{- !$disabled{'sm2'} && !$disabled{'sm2-precomp'} -}]` shouldn't it?

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/ec/asm/ecp_sm2p256-armv8.pl [hunk 18]
```diff
@@ -0,0 +1,820 @@
...
+sub bn_mod_sub() {
+	my $mod = shift;
+$code.=<<___;
+	# Load inputs
+	ldp $s0,$s1,[x1]
+	ldp $s2,$s3,[x1,#16]
+	ldp $s4,$s5,[x2]
+	ldp $s6,$s7,[x2,#16]
+
+	# Substraction
```
Comment (t8m) [inline crypto/ec/asm/ecp_sm2p256-armv8.pl:80]: Typo: Subtraction

File: crypto/ec/asm/ecp_sm2p256-armv8.pl [hunk 19]
```diff
@@ -0,0 +1,820 @@
...
+	sbcs $s2,$s2,$s6
+	sbcs $s3,$s3,$s7
+	sbc $t4,xzr,xzr
+
+	# Load polynomial
+	adr x2,$mod
+	ldp $s4,$s5,[x2]
+	ldp $s6,$s7,[x2,#16]
+
+	# Backup Substraction
```
Comment (t8m) [inline crypto/ec/asm/ecp_sm2p256-armv8.pl:92]: Typo: subtraction

File: crypto/ec/asm/ecp_sm2p256-armv8.pl [hunk 20]
```diff
@@ -0,0 +1,820 @@
...
+.align	5
+bn_sub:
+	AARCH64_VALID_CALL_TARGET
+	# Load inputs
+	ldp $s0,$s1,[x1]
+	ldp $s2,$s3,[x1,#16]
+	ldp $s4,$s5,[x2]
+	ldp $s6,$s7,[x2,#16]
+
+	# Substraction
```
Comment (t8m) [inline crypto/ec/asm/ecp_sm2p256-armv8.pl:212]: As above

**Review by t8m — COMMENTED** [review]
A few more nits

File: crypto/ec/build.info (hunk 17 above)
Comment (xu-yi-zhou) [reply crypto/ec/build.info:35]: In fact, SM2 algorithm is also available if there is no precomputed table. As shown in the following code, when there is no precomputed table, the program will use `ecp_sm2p256_windowed_mul()` to calculate the scalar multiplication of the base point.

```
#if !defined(OPENSSL_NO_SM2_PRECOMP)
        if (ecp_sm2p256_is_affine_G(generator)) {
            ecp_sm2p256_point_G_mul_by_scalar(&p.p, k);
        } else
#endif
        {
            /* if no precomputed table */
            const EC_POINT *new_generator[1];
            const BIGNUM *g_scalars[1];

            new_generator[0] = generator;
            g_scalars[0] = scalar;

            if (!ecp_sm2p256_windowed_mul(group, &p.p, g_scalars, new_generator,
                                          (new_generator[0] != NULL
                                           && g_scalars[0] != NULL), ctx))
                goto err;
        }
```

Comment (xu-yi-zhou) [reply crypto/ec/build.info:35]: I've updated the performance data for the no pre-computed table in the first comment.

**Review by t8m — APPROVED** [review]

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by sususu98** [comment]
> This patch optimizes SM2 for ARM processor using A64 instruction and precomputation table, which can speed up SM2 sign about 10 times and SM2 verify about 3 times. A new configure option `no-sm2-precomp` has been added to disable the precomputed table for point multiplicatin of the base point.
> 
> Perf data on Kunpeng-920 2.6GHz hardware looks like this:
> 
> sign	verify	sign/s	verify/s
> Before	0.0004s	0.0004s	2438.2	2460.9
> After	0.0000s	0.0001s	27137.4	7434.0
> After (no-sm2-precomp)	0.0001s	0.0002s	7745.5	4411.2
> Signed-off-by: Xu Yizhou [xuyizhou1@huawei.com](mailto:xuyizhou1@huawei.com)
> 
> ##### Checklist
> * [ ]  documentation is added or updated

Hello, I tested using the default compilation options on Kunpeng-920 2.6GHz hardware and obtained a signature performance of 25057 times/s and a signature verification performance of 7042 times/s. May I ask if you added other options to the data obtained during compilation

**Comment by xu-yi-zhou** [comment]
> Hello, I tested using the default compilation options on Kunpeng-920 2.6GHz hardware and obtained a signature performance of 25057 times/s and a signature verification performance of 7042 times/s. May I ask if you added other options to the data obtained during compilation

no other options, try using the `taskset `command to bind the CPU while testing?

**Pull request closed without merging**